### PR TITLE
distage testkit docs part 1

### PR DESCRIPTION
**Pull request by coreyoconnor**
For #820

### CODE REVIEW

File: doc/microsite/src/main/tut/distage/distage-testkit.md [hunk 1]
```diff
@@ -3,10 +3,265 @@ distage-testkit
...
 
-### Testkit
+### Quick Start
 
-[distage Example Project](https://github.com/7mind/distage-example) project shows how to use `distage-testkit`, there's also
-an overview in the [Hyper-pragmatic Pure FP Testing with distage-testkit](https://www.youtube.com/watch?v=CzpvjkUukAs) talk. 
+For effect types of `F[_]`, `F[+_, +_]` and `F[-_, +_, +_]` the `distage-testkit` library provides
+[ScalaTest](https://www.scalatest.org/) base classes. These provide an interface similar to
+ScalaTest's [`WordSpec`](http://doc.scalatest.org/3.1.0/org/scalatest/wordspec/AnyWordSpec.html).
+However, distage `SpecScalatest`s have additional capabilities. Such as first class support for
```
Comment (neko-kai) [inline doc/microsite/src/main/tut/distage/distage-testkit.md]:
```suggestion
However, distage `DistageSpecScalatest`s have additional capabilities. Such as first class support for
```

File: doc/microsite/src/main/tut/distage/distage-testkit.md [hunk 2]
```diff
@@ -3,10 +3,265 @@ distage-testkit
...
+        score.copy(value = value)
+    }
+
+  def addMango(score: Score): RIO[Has[BonusService] with Has[Config], Score] =
+    RIO.service[Config] flatMap { config => addMango(config, score) }
+}
+
+```
+
+#### `SpecScalatest` Base Classes
```
Comment (neko-kai) [inline doc/microsite/src/main/tut/distage/distage-testkit.md]:
```suggestion
#### `DistageSpecScalatest` Base Classes
```
I suppose it's a bit wierd to omit the prefix (even though there's a clash with the actual class name)

File: doc/microsite/src/main/tut/distage/distage-testkit.md [hunk 3]
```diff
@@ -3,10 +3,265 @@ distage-testkit
...
+For effect types of `F[_]`, `F[+_, +_]` and `F[-_, +_, +_]` the `distage-testkit` library provides
+[ScalaTest](https://www.scalatest.org/) base classes. These provide an interface similar to
+ScalaTest's [`WordSpec`](http://doc.scalatest.org/3.1.0/org/scalatest/wordspec/AnyWordSpec.html).
+However, distage `SpecScalatest`s have additional capabilities. Such as first class support for
+effect types and dependency injection.
+
+Usage of `distage-testkit` generally follows these steps:
+
+1. Select a base class corresponding to the effect type:
+  - `F[_]` - @scaladoc[`DistageSpecScalatest`](izumi.distage.testkit.scalatest.DistageSpecScalatest)
```
Comment (neko-kai) [inline doc/microsite/src/main/tut/distage/distage-testkit.md]: I think you may add the `No effect type / Identity` case here as well (which would be `DistageSpecScalatest[Identity]`)

File: doc/microsite/src/main/tut/distage/distage-testkit.md [hunk 4]
```diff
@@ -1,14 +1,518 @@
...
+    - @scaladoc[`in` for `F[+_, +_]`](izumi.distage.testkit.services.scalatest.dstest.DistageAbstractScalatestSpec$$DSWordSpecStringWrapper2)
+    - @scaladoc[`in` for `F[-_, +_, +_]`](izumi.distage.testkit.services.scalatest.dstest.DistageAbstractScalatestSpec$$DSWordSpecStringWrapper3)
+    - Magnet for test cases dependent on injectables: @scaladoc[`ProviderMagnet`](izumi.distage.model.providers.ProviderMagnet)
+
+### API Overview
+
+To demonstrate the usage of `distage-testkit` we'll consider a hypothetical game scoring system. This
+will have a model, logic and service which we'll then define test cases to verify.
+
+The highest value tests to develop, [in our experience](https://www.youtube.com/watch?v=CzpvjkUukAs),
```
Comment (neko-kai) [inline doc/microsite/src/main/tut/distage/distage-testkit.md]: Might better link to [blog post](https://blog.7mind.io/constructive-test-taxonomy.html) here or below when new terminology starts, and as the blog post is complimentary and blogs are more popular than videos.

File: doc/microsite/src/main/tut/distage/distage-testkit.md [hunk 5]
```diff
@@ -1,14 +1,518 @@
...
+is a function from `Config` the required argument will be injected by `distage-testkit`.
+
+```scala
+package app
+```
+```scala mdoc:to-string
+final class ScoreEffectsTest extends Test {
+  "testkit operations with effects" should {
+
+    "assertions in effects" in { (config: Config) =>
```
Comment (neko-kai) [inline doc/microsite/src/main/tut/distage/distage-testkit.md]:
```suggestion
    "assertions in effects" in {
       config: Config =>
```

Please keep arguments on a separate line

File: doc/microsite/src/main/tut/distage/distage-testkit.md [hunk 6]
```diff
@@ -1,14 +1,518 @@
...
+
+Our example application has a dummy and production implementation of the `BonusService`.  For each
+implementation a `ZManaged` is provided. With the `ZManaged` resources added to the object
+graph test cases can inject `Has[BonusService]`.
+
+For demonstration of [reuse and memoization](#resource-reuse-memoization) the bonus value will be
+equal to the number of times the resource was acquired.
+
+```scala
+package app
```
Comment (neko-kai) [inline doc/microsite/src/main/tut/distage/distage-testkit.md]: Are you planning on making the package significant in some way (e.g. plugins?) For plugins we may only need the plugin to be defined in a non-default package. (I think mdoc will not allow defining something in a package, but you may add any code you need to `doc/microsite/src/main/scala` and reference it in the docs, e.g. plugins)

**Review by neko-kai — APPROVED** [review]

File: doc/microsite/src/main/tut/distage/distage-testkit.md (hunk 4 above)
Comment (coreyoconnor) [reply doc/microsite/src/main/tut/distage/distage-testkit.md]: Ah! Thanks. Very useful. Any other references on testkit to include? I'll at least collect them in `references` if I can't find a way to integrate into the docs

File: doc/microsite/src/main/tut/distage/distage-testkit.md (hunk 6 above)
Comment (coreyoconnor) [reply doc/microsite/src/main/tut/distage/distage-testkit.md]: Right. Demonstrating plugin functionality is the goal. Specifically how the the default test config included the current package's plugins.
RE: mdoc not supporting defining something in a package. I encountered that as well. That's why the `package` statements are in plain scala blocks. A bit odd no? Still, somewhat captures the importance of packages.
 
You might have seen the note below related to running a test with the fake package: https://github.com/7mind/izumi/pull/1145/commits/32965b436e016edd4e9d540ece8aac60e83dabb3#diff-9cc6af0950016de37c81fb0f5609a6e6R403

I'll see about cleaning that up using `src/main/scala`. Was not aware of that feature. Thanks!

File: doc/microsite/src/main/tut/distage/distage-testkit.md [hunk 7]
```diff
@@ -1,14 +1,517 @@
...
+- `in { (a: A, b: B) => assert(???) }` - test case is a function returning an
+   assertion. The `a` and `b` will be injected from object graph.
+- `in { (a: A, b: B) => ???: F[Unit] }` - test case is a function returning an effect to be
+  executed. The `a` and `b` will be injected from object graph. The test case will fail if the effect
+  fails.
+- `in { (a: A, b: B) => ???: F[Assertion] }` - test case is a function returning an effect to be
+  executed. The `a` and `b` will be injected from object graph. The test case will fail if the effect
+  fails or produces a failure assertion.
+
+For `F[-_, +_, +_]`, same as above with `F[Any, Throwable, _]`:
```
Comment (neko-kai) [inline doc/microsite/src/main/tut/distage/distage-testkit.md]: @coreyoconnor  Note that using environment does not disable using arguments, the following is valid too:
```
`in { (a: A, b: B) => ???: F[zio.Has[C] with zio.Has[D], Throwable, Assertion] }`
```
All of A, B, C & D will be injected from the object graph

File: doc/microsite/src/main/tut/distage/distage-testkit.md (hunk 4 above)
Comment (neko-kai) [reply doc/microsite/src/main/tut/distage/distage-testkit.md]: @coreyoconnor 
I think that's the only blog post, but in release notes there were more detailed descriptions of some of the new testkit features:
* memoization environment merging - https://github.com/7mind/izumi/releases/tag/v0.10.9
* `TestConfig#forcedRoots`, `#parallelEnvs`, zio environment tests - https://github.com/7mind/izumi/releases/tag/0.10.2

Comment (neko-kai) [reply doc/microsite/src/main/tut/distage/distage-testkit.md]: Ah, slides for ["Hyper-pragmatic pure FP testing with distage-testkit"](https://www.slideshare.net/7mind/hyperpragmatic-pure-fp-testing-with-distagetestkit) & ["Scala, Functional Programming and Team Productivity"](https://www.slideshare.net/7mind/scala-functional-programming-and-team-productivity) both show examples of `distage-testkit` usage.

File: doc/microsite/src/main/tut/distage/distage-testkit.md [hunk 8]
```diff
@@ -1,14 +1,526 @@
...
+- `in { (a: A, b: B) => ???: F[Unit] }` - test case is a function returning an effect to be
+  executed. The `a` and `b` will be injected from the object graph. The test case will fail if the
+  effect fails.
+- `in { (a: A, b: B) => ???: F[Assertion] }` - test case is a function returning an effect to be
+  executed. The `a` and `b` will be injected from the object graph. The test case will fail if the
+  effect fails or produces a failure assertion.
+
+For `F[-_, +_, +_]`, same as above with `F[Any, Throwable, _]`:
+
+- `in { ???: F[zio.Has[C] with zio.Has[D], Throwable, Unit] }` - test case is an effect requiring an
```
Comment (neko-kai) [inline doc/microsite/src/main/tut/distage/distage-testkit.md]: @coreyoconnor Actually, another technical correction, for both `F[+_, +_]` the error type is `_`/`Any`, not `Throwable` (it's very convenient to not worry about conversions between error types in tests, so they are completely lenient)


Screenshot 2020-08-14 at 22 09 05

^ in all of the `in` constructors, the error type is wildcarded

File: doc/microsite/src/main/tut/distage/distage-testkit.md (hunk 6 above)
Comment (neko-kai) [reply doc/microsite/src/main/tut/distage/distage-testkit.md]: @coreyoconnor Hi, I've added a modifier to mdoc to allow writing a (fake) package statement in the same block - hope that helps! https://github.com/dogheadbone/izumi/pull/1

**Pull request merged**